### PR TITLE
Adding rendering views.

### PR DESCRIPTION
**Pull request by awais786**
ECOM-3445
Adding the back-end view code for rendering the template. For the time being we don't have any HTML so I just pass the values into the template. Also Code is written in way that it loads the relevant html file depending on the content type.

### CODE REVIEW

File: credentials/apps/credentials/views.py [hunk 1]
```diff
@@ -3,21 +3,110 @@
...
+                raise Http404
+
+            signatories = program.signatories.all()
+        else:
+            # Only Programs credentials are supported.
+            raise Http404
+
+        # if program has a template in db then it will render otherwise uses
+        # the default static template.
+        if program.template:
```
Comment (zubair-arbi) [inline credentials/apps/credentials/views.py]: this code should be part of `if generic_model == ProgramCertificate:`

**Comment by awais786** [comment]
@zubair-arbi please review this PR.

**Comment by zubair-arbi** [comment]
Lgtm :+1:

**Comment by awais786** [comment]
@jimabramson Please review this PR. I have created new PR for views.  
Following PR will be use to add the html stuff.
https://github.com/edx/credentials/pull/21

**Comment by awais786** [comment]
@clintonb Please review this PR. If you are busy please let me know.

**Comment by jimabramson** [comment]
@peter-fogg could you take my review here too.  (last one :-))

File: credentials/apps/credentials/views.py [hunk 2]
```diff
@@ -3,21 +3,116 @@
...
+        """ Get the program certificate related data from database."""
+        try:
+            program = ProgramCertificate.objects.get(pk=user_credential.credential_id)
+        except ProgramCertificate.DoesNotExist:
+            raise Http404
+
+        signatories = program.signatories.all()
+
+        # if program has a template in db then it will render otherwise uses
+        # the default static template.
```
Comment (peter-fogg) [inline credentials/apps/credentials/views.py]: I think this comment could be a little more clear -- maybe something like "If the program has a template in the database, use it for rendering. Otherwise, use the default static template."

File: credentials/apps/credentials/views.py [hunk 3]
```diff
@@ -3,21 +3,116 @@
...
+        """ Render the template from database."""
+        program_template = Template(program.template.content)
+        context = Context({
+            'content': program.template.content
+        })
+        context.update(self._update_certificate_context(user_credential))
+        program_certificate_context = program_template.render(context)
+        return program_certificate_context
+
+    def render_default_template(self, credential, signatories):
```
Comment (peter-fogg) [inline credentials/apps/credentials/views.py]: This function is very similar to `render_program_certificate` above -- could you create one function with the program-specific parts factored out?

File: credentials/apps/credentials/views.py [hunk 4]
```diff
@@ -3,21 +3,116 @@
...
+        program_certificate_context = program_template.render(context)
+        return program_certificate_context
+
+    def _update_certificate_context(self, user_certificate):
+        """
+        Helper method to build up the certificate web view context using the
+        provided values.
+        """
+
+        # Translators:  The format of the date includes the full name of the month.
```
Comment (peter-fogg) [inline credentials/apps/credentials/views.py]: Tiny nitpick: extra space after "Translators:".

File: credentials/apps/credentials/views.py [hunk 5]
```diff
@@ -3,21 +3,116 @@
...
+        certificate_date_issued = _('{month} {day}, {year}').format(
+            month=user_certificate.modified.strftime("%B"),
+            day=user_certificate.modified.day,
+            year=user_certificate.modified.year
+        )
+        # Translators: This text is bound to the HTML 'title' element of the page and appears in the browser title bar.
+        document_title = _('{partner_short_name} {course_number} Certificate | {platform_name}').format(
+            partner_short_name='edX',
+            course_number='XSeries',
```
Comment (peter-fogg) [inline credentials/apps/credentials/views.py]: Is `course_number` the right name for this variable? Seems like `certificate_type` or something might be more clear.

File: credentials/apps/credentials/views.py [hunk 6]
```diff
@@ -3,21 +3,116 @@
...
+        )
+        return {
+            'dir_rtl': 'rtl' if get_language_bidi() else 'ltr',
+            'LANGUAGE_CODE': self.request.LANGUAGE_CODE,
+            'platform_name': settings.PLATFORM_NAME,
+            'username': user_certificate.username,
+            'uuid': user_certificate.uuid,
+            'certificate_date_issued': certificate_date_issued,
+            'document_title': document_title,
+            'organization': {  # TODO will fix once retrieve orgs.
```
Comment (peter-fogg) [inline credentials/apps/credentials/views.py]: I don't have much context for this comment (though it may be more clear to others) -- what needs to happen for this code to be changed? Same comment on line 117.

File: credentials/templates/credentials/static/view_certificate.html [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    <meta charset="UTF-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+
+    <title>{{document_title}}</title>
+
+    {% if dir_rtl == 'ltr' %}
+        <!--<link rel="stylesheet" href="certificate_template_assets/1/edx-web-certificates-ltr.css" />-->
+    {% else %}
+        <!--<link rel="stylesheet" href="edx-web-certificates-rtl.css" />-->
```
Comment (peter-fogg) [inline credentials/templates/credentials/static/view_certificate.html]: Why are these commented out?

File: credentials/apps/credentials/views.py (hunk 6 above)
Comment (awais786) [inline credentials/apps/credentials/views.py]: Zubair PR is in review. Once that merges than I will use that to get the data.

File: credentials/templates/credentials/static/view_certificate.html (hunk 7 above)
Comment (awais786) [inline credentials/templates/credentials/static/view_certificate.html]: This story is more related with backend changes.  And html rendering will be done in other story. Also we don't have any css at this time.

**Comment by awais786** [comment]
@peter-fogg It's ready for another review.

File: credentials/apps/credentials/views.py [hunk 8]
```diff
@@ -3,21 +3,105 @@
...
-            'uuid': kwargs.get('uuid'),
+            'rendered_certificate': rendered_certificate,
         })
-
         return context
+
+    def _get_program_certificate_context(self, user_credential):
+        """ Get the program certificate related data from database."""
+        try:
+            program = ProgramCertificate.objects.get(pk=user_credential.credential_id)
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Please stop using `program` to refer to `program_certificate`. I've seen this on multiple PRs. You want your names to be as understandable as possible. Use `program_certificate` when referring to  `ProgramCertificate` objects.

File: credentials/apps/credentials/views.py [hunk 9]
```diff
@@ -3,21 +3,105 @@
...
+        except ProgramCertificate.DoesNotExist:
+            raise Http404
+
+        signatories = program.signatories.all()
+
+        # If the program has a template in the database, use it for rendering.
+        # Otherwise, use the default static template.
+
+        if program.template:
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: This is weird. There should not be multiple ways to declare a template. Either we use file-based templates or database templates, not both.

File: credentials/apps/credentials/views.py [hunk 10]
```diff
@@ -3,21 +3,105 @@
...
+            program_template = Template(program.template.content)
+            context = Context({
+                'content': program.template.content
+            })
+        else:
+            program_template = loader.get_template('credentials/static/view_certificate.html')
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: This file is poorly named:

1. It should not be nested in a `/static/` directory. 
2. What is `view_certificate`? Is that for all certificates? Program certificates?

Make sure it is easy for developers to understand what the file contains based on the file name.

File: credentials/apps/credentials/views.py [hunk 11]
```diff
@@ -3,21 +3,105 @@
...
+        return program_certificate_context
+
+    def _update_certificate_context(self, user_certificate):
+        """
+        Helper method to build up the certificate web view context using the
+        provided values.
+        """
+
+        # Translators: The format of the date includes the full name of the month.
+        certificate_date_issued = _('{month} {day}, {year}').format(
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Dates can be formatted in the template. https://docs.djangoproject.com/en/1.9/topics/i18n/formatting/#std:templatefilter-localize

File: credentials/apps/credentials/views.py (hunk 9 above)
Comment (awais786) [inline credentials/apps/credentials/views.py]: Are you talking about db template and default template existence or about the implementation ?

Comment (awais786) [inline credentials/apps/credentials/views.py]: @clintonb do you mean I should move the default.html into db also. Some thing like fixture data ?

File: credentials/apps/credentials/views.py (hunk 10 above)
Comment (awais786) [inline credentials/apps/credentials/views.py]: It's a default template that's why I used static name. And It will be used for programs certificates only due to the designing structure and background logo etc.

File: credentials/apps/credentials/views.py (hunk 11 above)
Comment (awais786) [inline credentials/apps/credentials/views.py]: I used the intentionally because this method will be used for both custom template and default template. And in case of custom template user just need to add a simple place holder e.g {{certificate_date_issued}}

**Comment by awais786** [comment]
@clintonb Please tell me what kind of approach you want for custom template and default template ?

**Comment by clintonb** [comment]
@awais786 we should limit the scope of templates. The current use case only requires the use of a single XSeries template. Until we have a use case for database templates, we should not work on that path. Doing so would be a waste of time given the functionality will not be used anytime soon.

File: credentials/apps/credentials/tests/test_views.py [hunk 12]
```diff
@@ -6,17 +6,76 @@
 
 from django.test import TestCase
 from django.core.urlresolvers import reverse
+from credentials.apps.api.tests import factories
```
Comment (tasawernawaz) [inline credentials/apps/credentials/tests/test_views.py]: add new line.

File: credentials/apps/credentials/tests/test_views.py [hunk 13]
```diff
@@ -6,17 +6,76 @@
...
+        path = reverse('credentials:render', kwargs={'uuid': self.user_credential.uuid.hex})
+        response = self.client.get(path)
+        self.assertEqual(response.status_code, 404)
+
+    def _assert_user_credential_template_data(self, response, user_credential):
+        """ Verify the default template has the data. """
+        self.assertContains(response, user_credential.username)
+        self.assertContains(response, user_credential.uuid)
+
+        issued_data = '{month} {day}, {year}'.format(
```
Comment (tasawernawaz) [inline credentials/apps/credentials/tests/test_views.py]: s/issued_data/issued_date

**Comment by awais786** [comment]
@clintonb its ready for the final review.

File: credentials/apps/credentials/views.py [hunk 14]
```diff
@@ -3,21 +3,68 @@
...
+    template_name = 'credentials/program_certificate.html'
 
     def get_context_data(self, **kwargs):
         context = super(RenderCredential, self).get_context_data(**kwargs)
 
-        context.update({
-            'uuid': kwargs.get('uuid'),
-        })
+        certificate_uuid = kwargs.get('uuid')
+        try:
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Use [`get_object_or_404`](https://docs.djangoproject.com/en/1.9/topics/http/shortcuts/#get-object-or-404).

File: credentials/apps/credentials/views.py [hunk 15]
```diff
@@ -3,21 +3,68 @@
...
+            user_credential = UserCredential.objects.get(
+                uuid=certificate_uuid, status=UserCredential.AWARDED
+            )
+        except UserCredential.DoesNotExist:
+            raise Http404
+
+        # get the model class according the credential content type.
+        generic_model = user_credential.credential_content_type.model_class()
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Move this filtering to the query above. No sense in getting back data from the DB that we aren't going to use.

File: credentials/apps/credentials/views.py [hunk 16]
```diff
@@ -3,21 +3,68 @@
...
+
+    def _get_program_certificate_context(self, user_credential):
+        """ Get the program certificate related data from database."""
+        try:
+            program_certificate = ProgramCertificate.objects.get(pk=user_credential.credential_id)
+        except ProgramCertificate.DoesNotExist:
+            raise Http404
+
+        # Translators: This text is bound to the HTML 'title' element of the page and appears in the browser title bar.
+        document_title = _('{partner_short_name} {certificate_type} Certificate | {platform_name}').format(
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Why is this here instead of in the template?

File: credentials/apps/credentials/views.py [hunk 17]
```diff
@@ -3,21 +3,68 @@
...
+            partner_short_name='edX',
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: These values should not be hard-coded.

File: credentials/apps/credentials/views.py [hunk 18]
```diff
@@ -3,21 +3,68 @@
...
+            platform_name=settings.PLATFORM_NAME,
+        )
+        return {
+            'signatories': program_certificate.signatories.all(),
+            'dir_rtl': 'rtl' if get_language_bidi() else 'ltr',
+            'LANGUAGE_CODE': self.request.LANGUAGE_CODE,
+            'platform_name': settings.PLATFORM_NAME,
+            'user_credential': user_credential,
+            'document_title': document_title,
+            'organization': {  # TODO will fix once retrieve orgs.
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: You at least need to stub out functions to get this data, and the course count.

File: credentials/apps/credentials/views.py [hunk 19]
```diff
@@ -3,21 +3,68 @@
...
+
+        # Translators: This text is bound to the HTML 'title' element of the page and appears in the browser title bar.
+        document_title = _('{partner_short_name} {certificate_type} Certificate | {platform_name}').format(
+            partner_short_name='edX',
+            certificate_type='XSeries',
+            platform_name=settings.PLATFORM_NAME,
+        )
+        return {
+            'signatories': program_certificate.signatories.all(),
+            'dir_rtl': 'rtl' if get_language_bidi() else 'ltr',
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Move this, `language_code`, and `platform_name` to a [context processor](https://docs.djangoproject.com/es/1.9/ref/templates/api/#writing-your-own-context-processors).

File: credentials/apps/credentials/views.py [hunk 20]
```diff
@@ -3,21 +3,68 @@
...
+            raise Http404
+
+        # Translators: This text is bound to the HTML 'title' element of the page and appears in the browser title bar.
+        document_title = _('{partner_short_name} {certificate_type} Certificate | {platform_name}').format(
+            partner_short_name='edX',
+            certificate_type='XSeries',
+            platform_name=settings.PLATFORM_NAME,
+        )
+        return {
+            'signatories': program_certificate.signatories.all(),
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Get this from the credential/certificate object.

File: credentials/apps/credentials/views.py [hunk 21]
```diff
@@ -3,21 +3,68 @@
...
+from credentials.apps.credentials.models import UserCredential, ProgramCertificate
+
 
 logger = logging.getLogger(__name__)
 
 
 class RenderCredential(TemplateView):
     """ Certificate rendering view."""
-    template_name = 'credentials/view_certificate.html'
+    template_name = 'credentials/program_certificate.html'
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: What about other credential types?

File: credentials/apps/credentials/views.py [hunk 22]
```diff
@@ -3,21 +3,68 @@
...
+            certificate_type='XSeries',
+            platform_name=settings.PLATFORM_NAME,
+        )
+        return {
+            'signatories': program_certificate.signatories.all(),
+            'dir_rtl': 'rtl' if get_language_bidi() else 'ltr',
+            'LANGUAGE_CODE': self.request.LANGUAGE_CODE,
+            'platform_name': settings.PLATFORM_NAME,
+            'user_credential': user_credential,
+            'document_title': document_title,
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: This is not actually used.

File: credentials/apps/credentials/tests/test_views.py [hunk 23]
```diff
@@ -7,16 +7,76 @@
...
-        """
-        Test certificate page accessibility. Page will return 200 if uuid is valid.
+    def setUp(self):
+        super(RenderCredentialPageTests, self).setUp()
+
+        self.program_certificate = factories.ProgramCertificateFactory.create(template=None)
+        self.user_credential = factories.UserCredentialFactory.create(
+            credential=self.program_certificate
+        )
+        self.username = 'test-user'
```
Comment (peter-fogg) [inline credentials/apps/credentials/tests/test_views.py]: This looks to be unused; remove it if so.

File: credentials/apps/credentials/views.py (hunk 17 above)
Comment (awais786) [inline credentials/apps/credentials/views.py]: @marcotuts would you please guide about these values ?

File: credentials/apps/credentials/views.py [hunk 24]
```diff
@@ -3,21 +3,66 @@
...
+        return {
+            'partner_short_name': 'edX',
+            'certificate_type': 'XSeries',
+            'signatories': user_credential.credential.signatories.all(),
+            'user_credential': user_credential,
+            'organization': self._get_organization_data(),
+            'program_course_count': self._get_programs_courses()
+        }
+
+    def _get_programs_courses(self):
```
Comment (zubair-arbi) [inline credentials/apps/credentials/views.py]: I have merged my PR for getting data from `programs` api, `organizations` api and `user` api.
Now you can use following functions to get data from above mentioned api's from `credentials/apps/credentials/utils.py`:
* get_program(program_id)
* get_organization(organization_key)
* get_user(username)

**Comment by awais786** [comment]
@peter-fogg It's ready for final review.

File: credentials/apps/credentials/views.py (hunk 15 above)
Comment (awais786) [inline credentials/apps/credentials/views.py]: No need to add crentential filter in main query. Here I am selecting html file for rendering purpose.

File: credentials/apps/credentials/views.py [hunk 25]
```diff
@@ -3,21 +3,87 @@
...
+            UserCredential,
+            uuid=kwargs.get('uuid'),
+            status=UserCredential.AWARDED
+        )
+
+        # get the model class according the credential content type.
+        # It will be use to call the appropriate method for rendering.
+        generic_model = user_credential.credential_content_type.model_class()
+        if generic_model == ProgramCertificate:
+            rendered_certificate = self._get_program_certificate_context(user_credential, context)
```
Comment (peter-fogg) [inline credentials/apps/credentials/views.py]: @clintonb's comment about filtering in the DB query here hasn't been addressed.

File: credentials/apps/core/context_processors.py [hunk 26]
```diff
@@ -5,5 +5,6 @@
 def core(_request):
     """ Site-wide context processor. """
     return {
-        'platform_name': settings.PLATFORM_NAME
+        'platform_name': settings.PLATFORM_NAME,
+        'language_code': _request.LANGUAGE_CODE,
```
Comment (peter-fogg) [inline credentials/apps/core/context_processors.py]: Change the `_request` parameter to `request` -- the underscore prefix is for unused variables.

File: credentials/apps/credentials/views.py (hunk 25 above)
Comment (awais786) [inline credentials/apps/credentials/views.py]: I have added the comment above also.
No need to add crentential filter in main query.

File: credentials/apps/core/tests/test_context_processors.py [hunk 27]
```diff
@@ -1,9 +1,11 @@
 """ Context processor tests. """
 
-from django.test import TestCase, override_settings, RequestFactory
+from django.test import override_settings, RequestFactory, TestCase
 
 from credentials.apps.core.context_processors import core
 
+
+LANGUAGE_CODE = 'us'
```
Comment (clintonb) [inline credentials/apps/core/tests/test_context_processors.py]: That's not a valid language code. Try `'en'`.

File: credentials/apps/core/tests/test_context_processors.py [hunk 28]
```diff
@@ -13,4 +15,7 @@ class CoreContextProcessorTests(TestCase):
     @override_settings(PLATFORM_NAME=PLATFORM_NAME)
     def test_core(self):
         request = RequestFactory().get('/')
-        self.assertDictEqual(core(request), {'platform_name': PLATFORM_NAME})
+        request.LANGUAGE_CODE = LANGUAGE_CODE
+        self.assertEqual(
```
Comment (clintonb) [inline credentials/apps/core/tests/test_context_processors.py]: `assertDictEqual`

File: credentials/apps/credentials/views.py [hunk 29]
```diff
@@ -3,21 +3,86 @@
...
+        # Update context with the signatories and user credential object
+        context.update({
+            'signatories': user_credential.credential.signatories.all(),
+            'user_credential': user_credential
+        })
+
+        program_template = loader.get_template('credentials/program_certificate.html')
+        return program_template.render(context, self.request)
+
+    def _get_programs_data(self, program_id):
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: You are only getting data fro a single program: `_get_program_data`.

File: credentials/apps/credentials/views.py [hunk 30]
```diff
@@ -3,21 +3,86 @@
...
+        Arguments:
+            program_id (int): Unique id of the program for retrieval
+
+        Returns:
+            dict, representing a parsed program data returned by the Programs service.
+        """
+        program_data = get_program(program_id)
+        return {
+            'course_count': len(program_data['course_codes']),
+            'organization_key': program_data['organizations'][0]['key'],
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: What happens if there are multiple organizations?

File: credentials/apps/credentials/views.py [hunk 31]
```diff
@@ -3,21 +3,86 @@
...
+            context (dict): Data containing basic information for template
+
+        Returns:
+            Rendered the program template with available context data.
+        """
+        # Update context with programs data
+        programs_data = self._get_programs_data(user_credential.credential.program_id)
+        context.update({'programs_data': programs_data})
+
+        # Update context with organization data
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: You don't need to comment simple operations. In general your functions and variables should be so well-named that your code doesn't need comments.

File: credentials/apps/credentials/views.py (hunk 30 above)
Comment (awais786) [inline credentials/apps/credentials/views.py]: We had discussed this here too.  For the time being there will be only one organization.  
@zubair-arbi would you please confirm ?

File: credentials/apps/credentials/views.py [hunk 32]
```diff
@@ -3,21 +3,86 @@
...
+            raise Http404
 
         context.update({
-            'uuid': kwargs.get('uuid'),
+            'rendered_certificate': rendered_certificate,
         })
 
         return context
+
+    def _get_program_certificate_context(self, user_credential, context):
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: This function is doing too much. It's name implies that it returns context data; however, it is actually rendering the entire template. Rather than do all of this, it would be preferable to simply return context and let the template do the rendering.

```python
def get_program_certificate_context(self, user_credential):
  return {
    'programs_data': ...,
    'organization_data': ...,
    'credential_template': 'credentials/program_certificate.html',
  }

# Note: user_credential should be added regardless of certificate type in get_context_data().
```

In your template:
```python
{% include credential_template %}
```

File: credentials/templates/credentials/program_certificate.html [hunk 33]
```diff
@@ -0,0 +1,25 @@
+{% extends 'base.html' %}
+{% load compress %}
+{% load i18n %}
+{% load staticfiles %}
+
+{% block title %}
+    {% blocktrans %}
+        XSeries Certificate | {{ platform_name }}
```
Comment (clintonb) [inline credentials/templates/credentials/program_certificate.html]: Only XSeries Certificate needs to be marked for translation.

File: credentials/templates/credentials/program_certificate.html [hunk 34]
```diff
@@ -0,0 +1,25 @@
+{% extends 'base.html' %}
```
Comment (clintonb) [inline credentials/templates/credentials/program_certificate.html]: This should extend a certificate base template that does not have all of the chrome (e.g. header, footer).

File: credentials/templates/credentials/render_credential.html [hunk 35]
```diff
@@ -0,0 +1,3 @@
+{% block content %}
```
Comment (clintonb) [inline credentials/templates/credentials/render_credential.html]: This should extend base.html.

File: credentials/apps/credentials/views.py (hunk 30 above)
Comment (zubair-arbi) [inline credentials/apps/credentials/views.py]: We will need @jimabramson and @marcotuts input here.
My understanding is that a Program will be linked to atleast one organization so for one organization this is fine but I am not sure how we will handle multiple organizations on certificate rendering also how often this will be the case that a program is linked to more than one organization.

File: credentials/templates/credentials/program_certificate.html (hunk 34 above)
Comment (awais786) [inline credentials/templates/credentials/program_certificate.html]: @clintonb I will add these base templates in a html related story.

File: credentials/templates/credentials/render_credential.html (hunk 35 above)
Comment (awais786) [inline credentials/templates/credentials/render_credential.html]: Will implement these changes in credential rendering story.

File: credentials/apps/credentials/views.py [hunk 36]
```diff
@@ -3,21 +3,73 @@
...
+            user_credential (User): UserCredential object against for a uuid
+
+        Returns:
+            Data containing program data, organization data and
+            program credential template.
+        """
+        programs_data = self._get_program_data(user_credential.credential.program_id)
+        return {
+            'programs_data': programs_data,
+            'organization_data': get_organization(programs_data.get('organization_key')),
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Use `programs_data['organization_key']` to avoid hiding issues where the key might be missing.

File: credentials/apps/credentials/views.py [hunk 37]
```diff
@@ -3,21 +3,73 @@
...
+            UserCredential,
+            uuid=kwargs.get('uuid'),
+            status=UserCredential.AWARDED
+        )
+
+        # get the model class according the credential content type.
+        # It will be use to call the appropriate method for rendering.
+        generic_model = user_credential.credential_content_type.model_class()
+        if generic_model == ProgramCertificate:
+            context.update(self.get_program_certificate_context(user_credential))
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Assign this to its own key, `certificate_context`, to avoid potentially overwriting the existing context.

File: credentials/apps/credentials/views.py [hunk 38]
```diff
@@ -3,21 +3,73 @@
...
         context = super(RenderCredential, self).get_context_data(**kwargs)
+        user_credential = get_object_or_404(
+            UserCredential,
+            uuid=kwargs.get('uuid'),
+            status=UserCredential.AWARDED
+        )
+
+        # get the model class according the credential content type.
+        # It will be use to call the appropriate method for rendering.
+        generic_model = user_credential.credential_content_type.model_class()
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: This can be cleaned to make it immediately clear that we are only dealing with program certificates:

```python
if user_credential.credential_content_type.model_class() != ProgramCertificate:
  raise Http404

context.update({
  'user_credential': user_credential,
  'certificate_context': self.get_program_certificate_context(user_credential),
})
```

File: credentials/apps/credentials/views.py [hunk 39]
```diff
@@ -3,21 +3,73 @@
...
+            'user_credential': user_credential
         })
 
         return context
+
+    def get_program_certificate_context(self, user_credential):
+        """ Get the program certificate related data from database.
+
+        Arguments:
+            user_credential (User): UserCredential object against for a uuid
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Clean this up.

File: credentials/apps/credentials/views.py [hunk 40]
```diff
@@ -3,21 +3,73 @@
...
+
+        Returns:
+            Data containing program data, organization data and
```
Comment (clintonb) [inline credentials/apps/credentials/views.py]: Clean this up.

**Comment by clintonb** [comment]
:+1: (after remaining comments addressed)
